File: sorting.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

void printArray(const string& msg, const vector<int>& arr) {
    cout << msg << ": ";
    for (int num : arr) {
        cout << num << " ";
    }
    cout << endl;
}

void bubbleSort(vector<int> arr) {
    cout << "\n--- BUBBLE SORT ---" << endl;
    int n = arr.size();
    bool swapped;
    for (int i = 0; i < n - 1; i++) {
        swapped = false;
        for (int j = 0; j < n - i - 1; j++) {
            if (arr[j] > arr[j + 1]) {
                swap(arr[j], arr[j + 1]);
                swapped = true;
            }
        }
        printArray("Iterasi ke-" + to_string(i + 1), arr);
        if (!swapped) break;
    }
    printArray("Hasil Akhir Bubble Sort", arr);
}

void selectionSort(vector<int> arr) {
    cout << "\n--- SELECTION SORT ---" << endl;
    int n = arr.size();
    for (int i = 0; i < n; i++) {
        int min_idx = i;
        for (int j = i + 1; j < n; j++) {
            if (arr[min_idx] > arr[j]) {
                min_idx = j;
            }
        }
        swap(arr[i], arr[min_idx]);
        printArray("Iterasi ke-" + to_string(i + 1), arr);
    }
    printArray("Hasil Akhir Selection Sort", arr);
}

void insertionSort(vector<int> arr) {
    cout << "\n--- INSERTION SORT ---" << endl;
    int n = arr.size();
    for (int i = 1; i < n; i++) {
        int key = arr[i];
        int j = i - 1;
        while (j >= 0 && arr[j] > key) {
            arr[j + 1] = arr[j];
            j = j - 1;
        }
        arr[j + 1] = key;
        printArray("Iterasi ke-" + to_string(i), arr);
    }
    printArray("Hasil Akhir Insertion Sort", arr);
}

int partition(vector<int>& arr, int low, int high) {
    int pivot = arr[high];
    int i = (low - 1);
    for (int j = low; j <= high - 1; j++) {
        if (arr[j] <= pivot) {
            i++;
            swap(arr[i], arr[j]);
        }
    }
    swap(arr[i + 1], arr[high]);
    return (i + 1);
}

void quickSortRecursive(vector<int>& arr, int low, int high) {
    if (low < high) {
        int pi = partition(arr, low, high);
        printArray("Proses Partisi (pivot " + to_string(arr[pi]) + ")", arr);
        quickSortRecursive(arr, low, pi - 1);
        quickSortRecursive(arr, pi + 1, high);
    }
}

void runQuickSort(vector<int> arr) {
    cout << "\n--- QUICK SORT ---" << endl;
    quickSortRecursive(arr, 0, arr.size() - 1);
    printArray("Hasil Akhir Quick Sort", arr);
}

void mergeSortRecursive(vector<int>& arr) {
    if (arr.size() > 1) {
        int mid = arr.size() / 2;
        vector<int> L(arr.begin(), arr.begin() + mid);
        vector<int> R(arr.begin() + mid, arr.end());

        mergeSortRecursive(L);
        mergeSortRecursive(R);

        int i = 0, j = 0, k = 0;
        while (i < L.size() && j < R.size()) {
            if (L[i] < R[j]) {
                arr[k] = L[i];
                i++;
            } else {
                arr[k] = R[j];
                j++;
            }
            k++;
        }

        while (i < L.size()) {
            arr[k] = L[i];
            i++;
            k++;
        }

        while (j < R.size()) {
            arr[k] = R[j];
            j++;
            k++;
        }
        printArray("Proses Merge", arr);
    }
}

void runMergeSort(vector<int> arr) {
    cout << "\n--- MERGE SORT ---" << endl;
    mergeSortRecursive(arr);
    printArray("Hasil Akhir Merge Sort", arr);
}

int main() {
    int n;
    cout << "Masukkan jumlah data: ";
    cin >> n;

    vector<int> data(n);
    for (int i = 0; i < n; i++) {
        cout << "Masukkan data ke-" << i + 1 << ": ";
        cin >> data[i];
    }

    printArray("\nData awal", data);

    bubbleSort(data);
    selectionSort(data);
    insertionSort(data);
    runQuickSort(data);
    runMergeSort(data);

    return 0;
}

File: sorting.py
def print_array(msg, arr):
    print(f"{msg}: {arr}")

def bubble_sort(arr):
    n = len(arr)
    data = arr.copy()
    print("\n--- BUBBLE SORT ---")
    for i in range(n):
        swapped = False
        for j in range(0, n-i-1):
            if data[j] > data[j+1]:
                data[j], data[j+1] = data[j+1], data[j]
                swapped = True
        print_array(f"Iterasi ke-{i+1}", data)
        if not swapped:
            break
    print_array("Hasil Akhir Bubble Sort", data)

def selection_sort(arr):
    n = len(arr)
    data = arr.copy()
    print("\n--- SELECTION SORT ---")
    for i in range(n):
        min_idx = i
        for j in range(i+1, n):
            if data[min_idx] > data[j]:
                min_idx = j
        data[i], data[min_idx] = data[min_idx], data[i]
        print_array(f"Iterasi ke-{i+1}", data)
    print_array("Hasil Akhir Selection Sort", data)

def insertion_sort(arr):
    data = arr.copy()
    print("\n--- INSERTION SORT ---")
    for i in range(1, len(data)):
        key = data[i]
        j = i-1
        while j >= 0 and key < data[j]:
            data[j+1] = data[j]
            j -= 1
        data[j+1] = key
        print_array(f"Iterasi ke-{i}", data)
    print_array("Hasil Akhir Insertion Sort", data)

def quick_sort_recursive(arr, low, high):
    if low < high:
        pi = partition(arr, low, high)
        print_array(f"Proses Partisi (pivot {arr[pi]})", arr)
        quick_sort_recursive(arr, low, pi-1)
        quick_sort_recursive(arr, pi+1, high)

def partition(arr, low, high):
    pivot = arr[high]
    i = low - 1
    for j in range(low, high):
        if arr[j] <= pivot:
            i = i + 1
            arr[i], arr[j] = arr[j], arr[i]
    arr[i+1], arr[high] = arr[high], arr[i+1]
    return i + 1

def run_quick_sort(arr):
    data = arr.copy()
    print("\n--- QUICK SORT ---")
    quick_sort_recursive(data, 0, len(data)-1)
    print_array("Hasil Akhir Quick Sort", data)

def merge_sort_recursive(arr):
    if len(arr) > 1:
        mid = len(arr)//2
        L = arr[:mid]
        R = arr[mid:]

        merge_sort_recursive(L)
        merge_sort_recursive(R)

        i = j = k = 0
        while i < len(L) and j < len(R):
            if L[i] < R[j]:
                arr[k] = L[i]
                i += 1
            else:
                arr[k] = R[j]
                j += 1
            k += 1

        while i < len(L):
            arr[k] = L[i]
            i += 1
            k += 1

        while j < len(R):
            arr[k] = R[j]
            j += 1
            k += 1
        print_array("Proses Merge", arr)

def run_merge_sort(arr):
    data = arr.copy()
    print("\n--- MERGE SORT ---")
    merge_sort_recursive(data)
    print_array("Hasil Akhir Merge Sort", data)

if __name__ == "__main__":
    try:
        n = int(input("Masukkan jumlah data: "))
        data_awal = []
        for i in range(n):
            angka = int(input(f"Masukkan data ke-{i+1}: "))
            data_awal.append(angka)
        
        print(f"\nData awal: {data_awal}")
        
        bubble_sort(data_awal)
        selection_sort(data_awal)
        insertion_sort(data_awal)
        run_quick_sort(data_awal)
        run_merge_sort(data_awal)
        
    except ValueError:
        print("Harap masukkan angka yang valid.")